crates/escrow/src/lib.rs

```diff
@@ -0,0 +1,602 @@
+        pub fn increase_amount(
+            origin: OriginFor<T>,
+            #[pallet::compact] amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let locked_balance = Self::locked_balance(&who);
+            let now = Self::current_height();
+
+            // value MUST be non-zero
+            ensure!(!amount.is_zero(), Error::<T>::InvalidAmount);
+
+            // lock MUST exist first
+            ensure!(!locked_balance.amount.is_zero(), Error::<T>::LockNotFound);
+
+            // lock MUST NOT be expired
+            ensure!(locked_balance.end > now, Error::<T>::LockHasExpired);
+
+            Self::deposit_for(&who, amount, Zero::zero())?;
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(<T as Config>::WeightInfo::increase_unlock_height())]
+        #[transactional]
+        pub fn increase_unlock_height(
+            origin: OriginFor<T>,
+            unlock_height: T::BlockNumber,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let locked_balance = Self::locked_balance(&who);
+            let now = Self::current_height();
+
+            // lock time is rounded down to weeks
+            let unlock_height = Self::round_height(unlock_height);
+
+            // lock MUST NOT be expired
+            ensure!(locked_balance.end > now, Error::<T>::LockHasExpired);
+
+            // lock amount MUST be non-zero
+            ensure!(!locked_balance.amount.is_zero(), Error::<T>::InvalidAmount);
+
+            // lock duration MUST increase
+            ensure!(unlock_height > locked_balance.end, Error::<T>::InvalidHeight);
+
+            // height MUST NOT be greater than max
+            let max_period = T::MaxPeriod::get();
+            let end_height = now + max_period;
+            ensure!(unlock_height <= end_height, Error::<T>::InvalidHeight);
+
+            Self::deposit_for(&who, Zero::zero(), unlock_height)?;
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(<T as Config>::WeightInfo::withdraw())]
+        #[transactional]
+        pub fn withdraw(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            Self::remove_lock(&who)?;
+            Ok(().into())
+        }
+    }
+}
+
+type DefaultPoint<T> = Point<BalanceOf<T>, <T as frame_system::Config>::BlockNumber>;
+type DefaultLockedBalance<T> = LockedBalance<BalanceOf<T>, <T as frame_system::Config>::BlockNumber>;
+
+// "Internal" functions, callable by code.
+impl<T: Config> Pallet<T> {
+    fn current_height() -> T::BlockNumber {
+        frame_system::Pallet::<T>::block_number()
+    }
+
+    fn round_height(height: T::BlockNumber) -> T::BlockNumber {
+        let span = T::Span::get();
+        (height / span) * span
+    }
+
+    fn checkpoint(who: &T::AccountId, old_locked: DefaultLockedBalance<T>, new_locked: DefaultLockedBalance<T>) {
```

Comment:
This function is a bit hard to follow, partly because of curve's naming standards. I would love to see this function in the spec, together with some explanation about exactly what it is supposed to achieve, and also when it is expected to be called. One thing in particular that I don't immediately understand, is `SlopeChanges`. I thought this was to account for additional amounts being locked when you already have some tokens locked, but then I don't understand how it is indexed only by the blocknumber (and not the account)

Reply:
This pallet is a prototype port of the [vyper contract](https://github.com/curvefi/curve-dao-contracts/blob/master/contracts/VotingEscrow.vy). In the interest of time please review their code / spec for now and we will optimize later.